### PR TITLE
Feature: Add AutoGen AgentChat Integration

### PR DESCRIPTION
**Pull request by lonestarx1**
## Summary
This PR integrates Tenuo with [Microsoft AutoGen](https://github.com/microsoft/autogen), enabling fine-grained access control and guardrails for AI agents. 

**Note on AutoGen Architecture:**
AutoGen operates in two modes: `autogen-agentchat` (high-level agent workflow) and `autogen-core` (event-driven architecture). This PR focuses exclusively on **AgentChat** to demonstrate immediate impact on common tool-use patterns. Support for `autogen-core` is planned for a follow-up PR.

It introduces a `GuardBuilder` API to wrap AgentChat tools with Tenuo warrants, enforcing constraints and cryptographic authorization (Proof-of-Possession) at runtime.

## Key Features

*   **Dual-Tier Protection**:
    *   **Tier 1 (Constraints-only)**: Lightweight validation using `GuardBuilder` to enforce input patterns without warrants.
    *   **Tier 2 (Warrant + PoP)**: Cryptographic enforcement using `BoundWarrant` to require valid signatures and proof-of-possession for every tool call.
*   **Streaming TOCTOU Protection**: Implements a "buffer-verify-emit" strategy (`guard_stream`) to prevent Time-of-Check Time-of-Use attacks in streaming tool calls.
*   **Flexible Argument Extraction**: Automatically handles various payload shapes (dictionaries, Pydantic models, dataclasses, positional/keyword args).
*   **Zero-Config Wrapper**: New `protect()` helper for quickly securing tools with minimal code.
*   **Async Support**: Full support for asynchronous tool execution.

### Core Integration
*   `tenuo-python/tenuo/autogen.py`: Implemented `GuardBuilder`, `guard_tool`, `guard_tools`, and `protect` APIs. Added logic for argument extraction and streaming verification.

### Tests
*   `tenuo-python/tests/test_autogen_integration.py`: Added comprehensive test suite covering:
    *   Tier 1 & Tier 2 enforcement
    *   Async tool execution
    *   Argument extraction edge cases
    *   Streaming TOCTOU scenarios (buffering, skipping, emitting)
    *   Integration invariants (monotonic attenuation, fail-closed defaults)

### Documentation & Examples
*   `docs/autogen.md`: Added integration guide and installation instructions.
*   **New Demos**:
    *   `tenuo-python/examples/autogen_demo_unprotected.py`: Baseline unprotected workflow.
    *   `tenuo-python/examples/autogen_demo_protected_tools.py`: Basic tool protection.
    *   `tenuo-python/examples/autogen_demo_protected_attenuation.py`: Per-agent attenuation.
    *   `tenuo-python/examples/autogen_demo_guardbuilder_tier1.py`: Tier 1 (constraints) demo.
    *   `tenuo-python/examples/autogen_demo_guardbuilder_tier2.py`: Tier 2 (warrant + PoP) demo.

## Verification
Run the new test suite:
pytest tenuo-python/tests/test_autogen_integration.py

### CODE REVIEW

File: tenuo-python/tenuo/bound_warrant.py [hunk 1]
```diff
@@ -253,11 +265,17 @@ def validate(self, tool: str, args: dict) -> ValidationResult:
...
-        success = self._warrant.authorize(tool=tool, args=args, signature=bytes(pop_signature))
+        try:
+            # 1. Sign
+            pop_signature = self._warrant.sign(self._key, tool, args)
+
+            # 2. Verify (calls Rust authorize)
+            success = self._warrant.authorize(
+                tool=tool, args=args, signature=bytes(pop_signature)
+            )
+        except Exception as exc:
```
Comment (aimable100) [inline tenuo-python/tenuo/bound_warrant.py]: Using a genereric exception type would make us lose alot of details (and difficult to debug). We need to make sure error types are propagated to the user with as much details as possible. So, we should explicitly define all exception types or better yet (what we had before), don't catch them. Let them propagate naturally.
 Looking at `_ensure_authorized_bound()`, if we don't do try/except, the errors will be handled well.
 
 Recommendation: let's not catch the exception here.

**Review by aimable100 — COMMENTED** [review]
Impressive implementation of the integration. Let's fix the bug in bound_warrant.py and add more tests. Added a few nits comments worth considering well. Good work!

File: tenuo-python/uv.lock
Comment (aimable100) [inline tenuo-python/uv.lock]: Let's remove lock files as they are environment-specific and shouldn't be in the repo

File: tenuo-python/tenuo/autogen.py [hunk 2]
```diff
@@ -0,0 +1,674 @@
...
+        *,
+        is_async: bool,
+    ) -> Any:
+        if self._on_denial == "raise":
+            raise exc
+        if self._on_denial == "log":
+            logger.warning(
+                "Authorization denied (%s): %s", type(exc).__name__, exc
+            )
+            result = fn(*args, **kwargs)
```
[the remark this answers is not in the document]
Comment (lonestarx1) [reply tenuo-python/tenuo/autogen.py]: Right. thanks for catching it.

**Review by aimable100 — APPROVED** [review]
